### PR TITLE
Add long-term memory via self-hosted supermemory server

### PR DESCRIPTION
**Pull request by kyh**
## What

Trial integration of [supermemory local](https://supermemory.ai/docs/self-hosting/overview) as the agent's long-term memory. New auto-discovered extension bundle at `apps/desktop/src/agent/memory/`:

- **`memory` tool** with `add` / `search` / `profile` / `list` / `forget` actions against the local server's Memory API (`/v3/documents`, `/v4/search`, `/v4/profile`, `/v4/memories`). All memories live under one `inteligir` container tag.
- **Recall priming**: `before_agent_start` injects the distilled user profile (static facts + recent context) as a hidden message — same pattern as the tasks extension's active-task priming. Served from a cache with background refresh so turns never stall on a cold HTTP call (first load waits ≤1.5s).
- **Dependency-free client** (`client.ts`): plain fetch, field-by-field response narrowing (no casts), endpoints verified against the `supermemory` SDK v4.24 type definitions. `SUPERMEMORY_BASE_URL` / `SUPERMEMORY_API_KEY` override the default `http://localhost:6767` target (localhost requests are auto-authenticated by the server, so no key plumbing needed by default).

## Trying it out

```bash
npx supermemory local   # installs + runs the server; prompts for one LLM
                        # provider key (~/.supermemory/env) used for extraction
```

The extension is **inert until the server is running**: the tool answers with setup guidance and the priming hook stays silent. No binaries are installed by the app.

## Notes from poking at the server (v0.0.2)

- Single binary (~190MB, Bun), data in `$SUPERMEMORY_DATA_DIR` (default `./.supermemory`), local embeddings via `Xenova/bge-base-en-v1.5` (~106MB one-time download from HF).
- Requires one of `OPENAI_API_KEY` / `ANTHROPIC_API_KEY` / `GEMINI_API_KEY` / `GROQ_API_KEY` at boot for memory extraction — worth noting our desktop auth is OpenAI **OAuth** via pi, so there's no API key we can hand it automatically.
- The server hardcodes its listener on `::`; it crashed in this (IPv6-less) CI sandbox, so I could not run a live end-to-end test here — behavior is covered by unit tests against SDK-verified shapes instead. Normal macOS/Linux desktops have IPv6 loopback, so this shouldn't affect real use, but **please verify locally** with the desktop app + a running server.

## Follow-ups if the experiment sticks

- Own the server lifecycle (install via bundle `setup()`, spawn/health-check/shutdown from main) instead of user-managed.
- Settings UI for enabling memory + entering the extraction-model key.
- Feed conversation turns to the server automatically (today the model decides what to `add`).

## Testing

- `pnpm typecheck && pnpm lint && pnpm test && pnpm knip && pnpm format && pnpm build` all green (14 new tests in `memory-extension.test.ts`).

https://claude.ai/code/session_01YQZbY4Wi4DtL5wRMcpBZtr

---
_Generated by [Claude Code](https://claude.ai/code/session_01YQZbY4Wi4DtL5wRMcpBZtr)_

---

> [!NOTE]
> **Medium Risk**
> New external dependency and user-data persistence path; failures are handled gracefully but stored memories and hidden profile priming affect agent behavior across sessions.
> 
> **Overview**
> Adds **long-term memory** for the desktop agent via an opt-in, user-run **supermemory** server (default `http://localhost:6767`, overridable with `SUPERMEMORY_BASE_URL` / `SUPERMEMORY_API_KEY`).
> 
> A new **`memory` extension** registers a curated **`memory` tool** (`add`, `search`, `profile`, `list`, `forget`) backed by a small fetch HTTP client scoped to the `inteligir` container tag, with defensive response parsing and user-facing setup guidance when the server is unreachable. On each turn, **`before_agent_start`** injects a hidden distilled user profile from a **cached background refresh** (one-time cold wait ≤1.5s) so turns do not block on HTTP.
> 
> **`AGENTS.md`** documents when to remember, search, forget, and how to behave if memory is down. **Unit tests** cover tool behavior, limit clamping, profile-cache edge cases, and client error classification.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f243017e50a45777026849c4d9e0588bde66296c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
## Summary by cubic
Adds self-hosted long-term memory to the desktop agent via a local `supermemory` server. Includes a `memory` tool, recall priming from a cached profile with a one-turn staleness tradeoff, AGENTS.md guidance, and a small fetch client that shows setup help when the server is unreachable.

- **Bug Fixes**
  - Recall priming spends the cold wait once even if the first profile request hangs.
  - Clears the cached profile only when the server is unreachable; treats request timeouts as transient and keeps the last good profile.
  - Clamps `limit` to an integer in 1–100 for `search`/`list` before hitting the server.
  - Unreachable-server guidance is contextual: local default shows `npx supermemory local`; custom `SUPERMEMORY_BASE_URL` advises checking the URL/server.
  - `add` fails loudly if the server returns 2xx without a document id (no false success).

- **Migration**
  - Run `npx supermemory local` and provide one LLM provider key at `~/.supermemory/env`.
  - Optionally set `SUPERMEMORY_BASE_URL` and `SUPERMEMORY_API_KEY` for a remote server.

<sup>Written for commit f243017e50a45777026849c4d9e0588bde66296c. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/kyh/inteligir/pull/342?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: apps/desktop/resources/agent/AGENTS.md [hunk 1]
```diff
@@ -40,6 +41,17 @@ Rules for raw system access:
...
+## Memory
+
+You have long-term memory across conversations via the `memory` tool. Use it like a good chief of staff uses a notebook — quietly and constantly.
+
+- When the user tells you something durable — a preference, a decision, a project, a person — store it right away with `add`. One concise fact per call. Don't ask permission to remember.
+- Before answering anything that leans on prior context ("what was that place I liked?"), `search` first instead of guessing.
+- Each conversation starts with a distilled profile of what you already know about the user. Trust it for orientation; `search` for specifics.
+- Never store secrets, passwords, API keys, or one-off trivia. Memory is for facts that are still true next week.
+- When the user says "forget that", use `forget` — immediately, no debate.
+- If memory is unavailable (the local server isn't running), carry on without it. Bring up setup only if the user asks about memory.
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/resources/agent/AGENTS.md:53]: Known limitation, deliberately out of scope for this trial PR: `seedFile` is seed-once by design (COPYFILE_EXCL — it must never clobber a user-edited AGENTS.md), so this section only reaches fresh installs. Same applies to every bundled-resource edit (skills included), not just this one. The real fix is a versioned seed/merge story for bundled resources, tracked as a follow-up if the memory experiment sticks. For local testing now: delete `~/.inteligir/AGENTS.md` to re-seed. The `memory` tool itself works regardless — its description carries the usage mechanics; AGENTS.md only adds behavioral policy.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/agent/memory/extension.ts [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+  return {
+    refresh,
+    /** Cached profile; only the first read waits (up to firstLoadWaitMs). */
+    async get(): Promise<MemoryProfile | null> {
+      const pending = refresh();
+      if (!coldWaitSpent) {
+        coldWaitSpent = true;
+        await Promise.race([pending, new Promise((r) => setTimeout(r, firstLoadWaitMs))]);
+      }
+      return cached;
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/agent/memory/extension.ts:224]: Deliberate tradeoff, now documented in the cache's docstring (df1a4d4): reads never wait after cold start because the alternative — awaiting the refresh on every `get()` — reintroduces the per-turn HTTP blocking flagged in the two earlier findings. The cost is a one-turn-stale snapshot exactly at the moment the server dies, which is benign: the profile holds durable user facts (preferences, projects, people), not availability state, and the very next turn primes from the cleared cache.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by kyh** [comment]
**Closing — superseded, and unmergeable against current `main`.**

Opened 2026-06-11. In the ~800 commits since, every mechanism this PR builds on was removed:

1. **All four files target paths that no longer exist.** `apps/desktop/src/agent/` is gone — the agent became its own package (`packages/agent/`) in the restructure, and bundled resources moved to `packages/agent/resources/agent/`.
2. **The registration mechanism was deliberately removed.** This adds an *auto-discovered* extension bundle; `packages/agent/src/bundles.ts` now says: *"The old auto-discovery used Vite's `import.meta.glob`, which ties the host library to one bundler; a static list is honest and portable."* Adding an extension is one folder + one line in that registry, enforced by a disk-drift test.
3. **The goal shipped, differently.** Long-term agent memory landed as the agent-instructions + memory loop (roadmap #1): `vault/AGENTS.md` is loaded into chat and delegation sessions. That keeps memory as plain markdown the user owns inside their own vault — consistent with the product's local-first premise — rather than requiring a self-hosted server as a runtime dependency.

Nothing here is a judgment on the idea. If a richer memory surface is wanted later, it should be a fresh bundle in `packages/agent/` against the static registry, and it should start from what the AGENTS.md loop already does rather than re-adding a parallel store.
